File: alica_engine/src/engine/modelmanagement/factories/SynchronisationFactory.cpp
#include "engine/modelmanagement/factories/SynchronisationFactory.h"
#include "engine/model/Synchronisation.h"
#include "engine/modelmanagement/Strings.h"

namespace alica
{
Synchronisation* SynchronisationFactory::create(const YAML::Node& synchronisationNode, Plan* plan)
{
    auto* synchronisation = new Synchronisation(plan, AlicaTime::milliseconds(Factory::getValue<int>(synchronisationNode, alica::Strings::talkTimeout)),
            AlicaTime::milliseconds(Factory::getValue<int>(synchronisationNode, alica::Strings::syncTimeout)),
            Factory::getValue<bool>(synchronisationNode, alica::Strings::failOnSyncTimeout));
    Factory::setAttributes(synchronisationNode, synchronisation);
    Factory::storeElement(synchronisation, alica::Strings::synchronisation);

    if (Factory::isValid(synchronisationNode[alica::Strings::inSync])) {
        const YAML::Node& transitions = synchronisationNode[alica::Strings::inSync];
        for (YAML::const_iterator it = transitions.begin(); it != transitions.end(); ++it) {
            Factory::synchTransitionReferences.push_back(std::pair<int64_t, int64_t>(synchronisation->getId(), Factory::getReferencedId(*it)));
        }
    }

    return synchronisation;
}

void SynchronisationFactory::attachReferences()
{
    // synchTransitionReferences
    for (std::pair<int64_t, int64_t> pairs : Factory::synchTransitionReferences) {
        Synchronisation* s = (Synchronisation*) Factory::getElement(pairs.first);
        Transition* t = (Transition*) Factory::getElement(pairs.second);
        s->addInSync(t);
    }
    Factory::synchTransitionReferences.clear();
}
} // namespace alica

File: alica-supplementary/plan-conversion-tool/src/main/java/com/rapyutarobotics/alica/factories/SynchronisationFactory.java
package com.rapyutarobotics.alica.factories;

import com.rapyutarobotics.alica.ConversionProcess;
import com.rapyutarobotics.alica.Tags;
import de.unikassel.vs.alica.planDesigner.alicamodel.Synchronisation;
import de.unikassel.vs.alica.planDesigner.alicamodel.Transition;
import javafx.util.Pair;
import org.w3c.dom.Element;

public class SynchronisationFactory extends Factory {

    public static Synchronisation create(Element synchronisationNode, ConversionProcess cp) {
        Synchronisation synchronisation = new Synchronisation();
        Factory.setAttributes(synchronisationNode, synchronisation);
        cp.addElement(synchronisation);
        synchronisation.setFailOnSyncTimeout(Boolean.parseBoolean(synchronisationNode.getAttribute(Tags.FAILONSYNCTIMEOUT)));
        synchronisation.setSyncTimeout(Integer.parseInt(synchronisationNode.getAttribute(Tags.SYNCTIMEOUT)));
        synchronisation.setTalkTimeout(Integer.parseInt(synchronisationNode.getAttribute(Tags.TALKTIMEOUT)));

        String synchedTransitionsString = synchronisationNode.getAttribute(Tags.SYNCHEDTRANSITIONS);
        String[] transitionIdStrings = synchedTransitionsString.split(" ");
        for (String transitionIdString : transitionIdStrings) {
            cp.synchTransitionReferences.put(synchronisation.getId(), cp.getReferencedId(transitionIdString));
        }

        return synchronisation;
    }

    public static void attachReferences(ConversionProcess cp) {
        for (Pair<Long, Long> entry : cp.synchTransitionReferences.getEntries()) {
            Synchronisation synchronisation = (Synchronisation) cp.getElement(entry.getKey());
            Transition transition = (Transition) cp.getElement(entry.getValue());
            if (transition == null) {
                throw new RuntimeException("[SynchronisationFactory] Transition with ID " + entry.getValue() + " unknown!");
            }
            synchronisation.addSyncedTransition(transition);
        }
        cp.synchTransitionReferences.clear();
    }
}
